File: backend/app/api/routes/certificates.py
"""Certificate endpoints - behavioral ports of the PHP certificate APIs.

GET  /certificates            list own certificates
GET  /certificates?id=        single own certificate
POST /certificates            request a certificate for a qualifying attempt
GET  /verify-certificate?id=  PUBLIC verification (no auth, public-safe fields)

Certificate PDF "download" is client-side in the existing frontend (jsPDF +
QRCode.js on the view page) - there is no server-side download endpoint in
PHP, so none is added here; the data contract it consumes is preserved.
"""

import re

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from app.api.deps import authed_with_token, require_csrf
from app.core.errors import APIError
from app.repositories.certificate_repository import CertificateRepository
from app.repositories.profile_repository import ProfileRepository
from app.repositories.quiz_repository import QuizRepository
from app.services.certificate_service import CertificateService, MIN_PERCENT
from app.utils.supabase import get_supabase

router = APIRouter(tags=["certificates"])

UUID_RE = re.compile(r"^[0-9a-fA-F-]{36}$")


@router.get("/certificates")
async def certificates(request: Request, id: str = "") -> dict:
    session, token = await authed_with_token(request)
    repo = CertificateRepository(get_supabase())

    if id:
        if not UUID_RE.match(id):
            raise APIError("Invalid certificate id.", 400, "bad_request")
        row = await repo.find_for_user(session.user["id"], id, token)
        if row is None:
            raise APIError("Certificate not found.", 404, "not_found")
        return {"error": None, "certificate": row}

    rows = await repo.list_for_user(session.user["id"], token)
    return {"error": None, "certificates": rows}


@router.post("/certificates")
async def create_certificate(request: Request) -> dict:
    session, token = await authed_with_token(request)
    body = await request.json() if await request.body() else {}
    if not isinstance(body, dict):
        body = {}
    require_csrf(request, session, body)

    attempt_id = str(body.get("attempt_id") or "")
    if not UUID_RE.match(attempt_id):
        raise APIError("A valid attempt id is required.", 400, "bad_request")

    db = get_supabase()
    quiz_repo = QuizRepository(db)
    found = await quiz_repo.find_attempt_for_user(session.user["id"], attempt_id, token)
    if found is None:
        raise APIError("That attempt could not be found.", 404, "not_found")

    attempt = found["attempt"]
    percent = float(attempt.get("percent") or 0)
    quiz_id = str(attempt.get("quiz_id") or "")

    quiz_title = "Completed Quiz"
    if quiz_id:
        quiz = await quiz_repo.find_quiz_for_user(session.user["id"], quiz_id, token)
        if quiz:
            title = str(quiz["quiz"].get("title") or "").strip()
            if title:
                quiz_title = title

    profile = await ProfileRepository(db).get(session.user["id"], token)
    name = str(
        (profile or {}).get("full_name")
        or session.user.get("full_name")
        or "Learner"
    )

    created = await CertificateService(db).create_for_attempt(
        session.user["id"], token, attempt_id, quiz_id, quiz_title, percent, name
    )
    if created is None:
        raise APIError(
            f"Certificates are awarded for scores of {int(MIN_PERCENT)}% or higher on a completed quiz.",
            422,
            "ineligible",
        )
    return JSONResponse(status_code=201, content={"error": None, "certificate": created})


@router.get("/verify-certificate")
async def verify_certificate(request: Request, id: str = "") -> dict:
    """Public verification - no auth; returns only public-safe fields."""
    if not UUID_RE.match(id or ""):
        raise APIError("Invalid certificate id.", 400, "bad_request")

    row = await CertificateRepository(get_supabase()).verify_public(id)
    if row is None:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "verified": False,
                "message": "This certificate could not be verified.",
            },
        )

    return {"error": None, "verified": True, "certificate": row}

File: api/certificates.php
<?php
/**
 * QuizSphere - Certificates API
 * -------------------------------------------------------------
 * GET  (auth)            -> list the user's certificates, or a single one ?id=
 * POST (auth, CSRF)      -> request a certificate for a specific quiz attempt.
 *
 * Eligibility (>=80%) and the data shown are always derived server-side from
 * the persisted attempt — never trusted from the client.
 */

declare(strict_types=1);

require_once __DIR__ . '/helpers.php';

if ($_SERVER['REQUEST_METHOD'] !== 'GET' && $_SERVER['REQUEST_METHOD'] !== 'POST') {
    Security::json(['error' => 'method_not_allowed', 'message' => 'Use GET or POST.'], 405);
}

$services = api_services();
if (!$services) {
    exit;
}
['client' => $client, 'auth' => $auth, 'repo' => $repo] = $services;

api_require_auth($auth);
$userId = $auth->id();
$token  = $auth->accessToken();
if (!$token) {
    Security::json(['error' => 'unauthenticated', 'message' => 'Your session expired.'], 401);
}

$cert = new CertificateService($client);

// ---------------------------------------------------------------
// POST: request a certificate for a completed attempt.
// ---------------------------------------------------------------
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $input = Security::readJsonInput();
    api_require_csrf($input);

    $attemptId = (string) ($input['attempt_id'] ?? '');
    if (!preg_match('/^[0-9a-fA-F-]{36}$/', $attemptId)) {
        Security::json(['error' => 'bad_request', 'message' => 'A valid attempt id is required.'], 400);
    }

    try {
        $found = $repo->findAttemptForUser($userId, $attemptId, $token);
    } catch (SupabaseException $e) {
        $found = null;
    }
    if (!$found) {
        Security::json(['error' => 'not_found', 'message' => 'That attempt could not be found.'], 404);
    }

    $attempt = $found['attempt'] ?? [];
    $percent = (float) ($attempt['percent'] ?? 0);
    $quizId  = (string) ($attempt['quiz_id'] ?? '');

    // Resolve quiz title server-side.
    $quizTitle = 'Completed Quiz';
    if ($quizId !== '') {
        $quiz = $repo->findQuizForUser($userId, $quizId, $token);
        if ($quiz) {
            $quizTitle = trim((string) ($quiz['quiz']['title'] ?? '')) !== '' ? $quiz['quiz']['title'] : $quizTitle;
        }
    }

    $name = (string) ($auth->profile()['full_name'] ?? ($auth->user()['full_name'] ?? 'Learner'));
    $created = $cert->createForAttempt($userId, $token, $attemptId, $quizId, $quizTitle, $percent, $name);

    if ($created === null) {
        Security::json([
            'error'   => 'ineligible',
            'message' => 'Certificates are awarded for scores of ' . CertificateService::MIN_PERCENT . '% or higher on a completed quiz.',
        ], 422);
    }

    Security::json(['error' => null, 'certificate' => $created], 201);
}

// ---------------------------------------------------------------
// GET: single by ?id=, else list.
// ---------------------------------------------------------------
$id = (string) ($_GET['id'] ?? '');
if ($id !== '') {
    if (!preg_match('/^[0-9a-fA-F-]{36}$/', $id)) {
        Security::json(['error' => 'bad_request', 'message' => 'Invalid certificate id.'], 400);
    }
    $row = $cert->findForUser($userId, $id, $token);
    if (!$row) {
        Security::json(['error' => 'not_found', 'message' => 'Certificate not found.'], 404);
    }
    Security::json(['error' => null, 'certificate' => $row], 200);
}

$list = $cert->listForUser($userId, $token);
Security::json(['error' => null, 'certificates' => $list], 200);
